tests/unit/test_symmetric_crypto.cpp

```diff
@@ -532,3 +536,62 @@ TEST_F(SymmetricAuthenticatedCipherTest, throwsWhenWrongTagIsUsed) {
 
     ASSERT_THROW(decryptor->finish(), MoCOCrWException);
 }
+
+TEST_F(SymmetricAuthenticatedCipherTest, authenticatesAssociatedData) {
+    auto encryptor = AESCipherBuilder{SymmetricCipherMode::GCM,
+                                      SymmetricCipherKeySize::S_256,
+                                      _secretKey}.buildAuthenticatedEncryptor();
+
+    encryptor->setAssociatedData(_associatedData);
+    encryptor->update(_plaintext);
+    auto ciphertext = encryptor->finish();
+    auto tag = encryptor->getAuthTag();
+
+    auto decryptor = AESCipherBuilder{SymmetricCipherMode::GCM,
+                                      SymmetricCipherKeySize::S_256,
+                                      _secretKey}.setIV(encryptor->getIV()).buildAuthenticatedDecryptor();
+
+    decryptor->setAssociatedData(_associatedData);
+    decryptor->update(ciphertext);
+    decryptor->setAuthTag(tag);
+    auto decryptedText = decryptor->finish();
+    ASSERT_THAT(decryptedText, ::testing::ElementsAreArray(_plaintext));
+}
+
+TEST_F(SymmetricAuthenticatedCipherTest, throwsIfSetAssociatedDataCalledAfterUpdate) {
+    auto encryptor = AESCipherBuilder{SymmetricCipherMode::GCM,
+                                      SymmetricCipherKeySize::S_256,
+                                      _secretKey}.buildAuthenticatedEncryptor();
+
+    encryptor->update(_plaintext);
+    ASSERT_THROW(encryptor->setAssociatedData(_associatedData), MoCOCrWException);
+}
+
+TEST_F(SymmetricAuthenticatedCipherTest, throwsIfAssociatedDataMissingOrModified) {
+    auto encryptor = AESCipherBuilder{SymmetricCipherMode::GCM,
+                                      SymmetricCipherKeySize::S_256,
+                                      _secretKey}.buildAuthenticatedEncryptor();
+
+    encryptor->setAssociatedData(_associatedData);
+    encryptor->update(_plaintext);
+    auto ciphertext = encryptor->finish();
+    auto tag = encryptor->getAuthTag();
+
+    auto decryptor = AESCipherBuilder{SymmetricCipherMode::GCM,
+                                      SymmetricCipherKeySize::S_256,
+                                      _secretKey}.setIV(encryptor->getIV()).buildAuthenticatedDecryptor();
+    decryptor->setAuthTag(tag);
+    _associatedData[2] ^= 1;
+    decryptor->setAssociatedData(_associatedData);
+    decryptor->update(ciphertext);
+
+    ASSERT_THROW(decryptor->finish(), MoCOCrWException);
+
+    auto decryptor2 = AESCipherBuilder{SymmetricCipherMode::GCM,
+                                      SymmetricCipherKeySize::S_256,
+                                      _secretKey}.setIV(encryptor->getIV()).buildAuthenticatedDecryptor();
+    decryptor2->setAuthTag(tag);
+    decryptor2->update(ciphertext);
+
+    ASSERT_THROW(decryptor2->finish(), MoCOCrWException);
+}
```

Comment:
Due to the slightly weird mechanism of OpenSSL to use the UpdateCipher method that is also used for encryption, I would appreciate a test that checks if the resulting ciphertext (without the auth tag) matches if encrypted with and without additional data. This way we will immediately see what is going on if something is subtly broken around this UpdateCipher semantics. Just encrypt the data twice, once with and once without additional data and check if it matches. Please note that you need to explicitly set the IV for this also on the Encryptor.

Reply:
Added a test